README.md
zsnapd
======

ZFS Snapshot Daemon

A rework of ZFS Snapshot Manager by Kenneth Henderick <kenneth@ketronic.be>

ZFS dataset configuration file /etc/zfssnapmanager.cfg should be upwards compatible with /etc/zsnapd/dataset.conf.

Usage
-----

All the functional code is in the scripts folder.  Execute zsnapd, it deamonizes itself.

Features
--------

* Fully Python3 based
* Laptop friendly as has built in connectivity test to check remote port reachability
* Remote mode - snapshotting, script execution, and snapshot aging from central backup server.
* Native Systemd daemon compitability via py-magcode-core python daemon and logging support library
* Debug command line switch and stderr logging
* Systemd journalctl logging.
* Full standard Unix daemon support via py-magcode-core, with logging to syslog or logfile
* Configuration is stored in configuration files with the ini file format.  There is a template file, and a dataset file.
* Triggers the configured actions based on time or a '.trigger' file present in the dataset's mountpoint, location read
  from /proc/mounts.
* Can take snapshots (with a yyyymmdd timestamp format)
* Can replicate snapshots to/from other nodes
  * Push based when the replication source has access to the replication target
  * Pull based when the replication source has no access to the replication target. Typically when you don't want to give
    all nodes access to the backup/replication target
* Full clone replication mode, to copy sub folders and all ZFS properties
* Cleans all snapshots with the yyyymmdd timestamp format based on a GFS schema (Grandfather, Father, Son).
* Supports pre and post commands
  * Pre command is executed before any action is executed
  * Post command is executed after the actions are executed, but before cleaning
* Has sshd remote execution filter script zsnapd-rcmd.  See etc/zsnapd/zsnapd-rcmd.conf for configuration.

Configuration
-------------

The daemon configuration file is in /etc/zsnapd/process.conf in ini format and a sample is as follows:

		[DEFAULT]
		run_as_user = root

		[zsnapd]
		# Use following setting to check daemon reconfiguring
		daemon_canary = blau
		debug_mark = True
		# Both below in seconds
		sleep_time = 300
		debug_sleep_time = 15
    # dataset configuration file
    # dataset_config_file = /etc/zsnapd/datasets.conf
    # dataset_config_file = /etc/zfssnapmanager.cfg
		# Uncomment to set up syslog logging
		# see pydoc3 syslog and man 3 syslog for value names with 'LOG_'
		# prefix stripped
		#syslog_facility = DAEMON
		#syslog_level = INFO
		# Uncomment to set up file logging
		#log_file = /var/log/zsnapd.log
		#log_file_max_size_kbytes = 1024
		#log_file_backup_count = 7

    [zsnapd-cfgtest]
    log_level = DEBUG

    [zsnapd-trigger]
    log_level = DEBUG

Adjust sleep_time (in seconds) to set interval zsnapd runs code.  For 30 minute intervals, set to
1800 seconds.

Command line arguments to zsnapd are:

		Usage: zsnapd [-dhv] [-c config_file]

			ZFS Snap Managment Daemon

			-c, --config-file       set configuration file
			-d, --debug             set debug level {0-3|none|normal|verbose|extreme}
			-h, --help              help message
			-b, --memory-debug      memory debug output
			-S, --systemd           Run as a systemd daemon, no fork
			-v, --verbose           verbose output
			-r, --rpdb2-wait        Wait for rpdb2 (seconds)

Note the default configuration file is /etc/zsnapd/process.conf, and systemd native mode is via the
--systemd switch

The dataset configuration file is located in /etc/zsnapd and is called dataset.conf. It's an .ini
file containing a section per dataset/volume that needs to be managed.  There is also a template .ini
file called template.conf in the same directory

zsnapd-cfgtest tests the data set conifugration file, and zsnapd-trigger writes out .trigger files
based on the data set configuration. It takes either the mount point of the target dataset as an argument, 
or the full dataset name including storage pool. zsnapd-trigger can optionally do a connectivity test first
before writing out the .trigger file. The same connectivty test is done before zsnapd attempts replication, and
it uses the replicate_endpoint_host and replicate_endpoint_port settings for the dataset. 

Examples

/etc/zsnapd/template.conf:

    [DEFAULT]
    replicate_endpoint_host = nas.local

    [backup-local]
    replicate_endpoint_port = 2345
    replicate_endpoint_command = ssh -l backup -p {port} {host}
    compression = gzip
    time = 17:00 - 21:00 /2

    [backup-other]
    replicate_endpoint_host = other.remote.server.org

/etc/zsnapd/dataset.conf:
    [DEFAULT]
    time = trigger

    [zroot]
    template = backup-local
    mountpoint = /
    time = {template}, trigger
    snapshot = True
    schema = 7d3w11m5y

    [zpool/data]
    mountpoint = /mnt/data
    time = trigger
    snapshot = True
    schema = 5d0w0m0y
    preexec = echo "starting" | mail somebody@example.com
    postexec = echo "finished" | mail somebody@exampe.com

    [zpool/myzvol]
    mountpoint = None
    time = 21:00
    snapshot = True
    schema = 7d3w0m0y

    [zpool/backups/data]
    template = backup-other
    mountpoint = /mnt/backups/data
    time = 08:00, 19:00-23:00/00:30
    snapshot = False
    replicate_source = zpool/data
    schema = 7d3w11m4y
    replicate_full_clone = True
    buffer_size = 128M

A summary of the different options:

* mountpoint: Points to the location to which the dataset is mounted, None for volumes.  Used only for triggers. Defaults to value in /proc/mounts if available.
* do_trigger: Dataset is a candidate for devops triggers with 'zsnapd-trigger -t'.
* time: Can be either a timestamp in 24h hh:mm notation after which a snapshot needs to be taken, a time range, 'trigger', or a comma separated list of such items. A time range consists of a start time, then a dash, an end time, with an optional interval separated by a '/'. The interval can be given in hours, or HH:MM format.  If not given, it is 1 hour.  Alone with the timestamps, 'trigger' indicates that it will take a snapshot as soon as a file with name '.trigger' is found in the dataset's mountpoint. This can be used in case data is for example rsynced to the dataset. As shown above, '{template}' is substituted for the time specification string from the template for that dataset in the dataset file.  Thus the time setting for an individual dataset using a template can be augmented in its definition.
* snapshot: Indicates whether a snapshot should be taken or not. It might be possible that only cleaning needs to be executed if this dataset is actually a replication target for another machine.
* replicate_append_basename: Append last part of dataset name (after last '/') to target dataset name and receive mount point, joining with a '/'.
* replicate2_append_basename: Append last part of dataset name (after last '/') to target dataset name and receive mount point, joining with a '/'.
* replicate_append_fullname: Append dataset name without pool name to target datset name and receive mount point, joining with a '/'.
* replicate2_append_fullname: Append dataset name without pool name to target datset name and receive mount point, joining with a '/'.
* replicate_endpoint: Deprecated. Can be left empty if replicating on localhost (e.g. copying snapshots to other pool). Should be omitted if no replication is required.
* replicate_endpoint_host: Can be left empty if replicating on localhost (e.g. copying snapshots to other pool). Should be omitted if no replication is required.
* replicate2_endpoint_host: Can be left empty if replicating on localhost (e.g. copying snapshots to other pool). Should be omitted if no replication is required.
* replicate_endpoint_port: port that has to be remotely accessed
* replicate2_endpoint_port: port that has to be remotely accessed
* replicate_endpoint_command: Command template for remote access. Takes two keys {port} and {host}
* replicate2_endpoint_command: Command template for remote access. Takes two keys {port} and {host}
* replicate_target: The target to which the snapshots should be send. Should be omitted if no replication is required or a replication_source is specified.
* replicate2_target: The target to which the snapshots should be send. Should be omitted if no replication is required or a replication_source is specified.
* replicate_source: The source from which to pull the snapshots to receive onto the local dataset. Should be omitted if no replication is required or a replication_target is specified.
* replicate_full_clone: Full clone of dataset and all sub ordinate datasets and properties
* replicate2_full_clone: Full clone of dataset and all sub ordinate datasets and properties
* replicate_receive_save: If transfer fails create a save point for resuming transfer
* replicate2_receive_save: If transfer fails create a save point for resuming transfer
* replicate_receive_mountpoint: Specify mount point value for received dataset (on remote if pushed).
* replicate2_receive_mountpoint: Specify mount point value for received remote dataset.
* replicate_receive_no_mountpoint: Remove mountpoint from received properties. Defaults to True if replicate_send_properties or replicate_full_clone is set.
* replicate_receive_no_mountpoint: Remove mountpoint from received properties. Defaults to True if replicate_send_properties or replicate_full_clone is set.
* replicate2_receive_no_mountpoint: Remove mountpoint from received properties. Defaults to True if replicate_send_properties or replicate_full_clone is set.
* replicate_receive_umount: Don't mount received dataset. Defaults to True if replicate_send_properties or replicate_full_clone is set.
* replicate2_receive_umount: Don't mount received dataset. Defaults to True if replicate_send_properties or replicate_full_clone is set.
* replicate_send_compression: zfs send using compressed data from disk
* replicate2_send_compression: zfs send using compressed data from disk
* replicate_send_raw: zfs send using raw data from disk
* replicate2_send_raw: zfs send using raw data from disk
* replicate_send_properties: zfs send sends all properties of dataset
* replicate2_send_properties: zfs send sends all properties of dataset
* replicate_full_clone: Full clone of dataset and all sub ordinate datasets and properties
* replicate2_full_clone: Full clone of dataset and all sub ordinate datasets and properties
* buffer_size: Give mbuffer buffer size in units of k, M, and G - kilobytes, Megabytes, and Gigabytes respectively.
* buffer2_size: Give mbuffer buffer size in units of k, M, and G - kilobytes, Megabytes, and Gigabytes respectively.
* compression: Indicates the compression program to pipe remote replicated snapshots through (for use in low-bandwidth setups.) The compression utility should accept standard compression flags (`-c` for standard output, `-d` for decompress.)
* compression2: Indicates the compression program to pipe remote replicated snapshots through (for use in low-bandwidth setups.) The compression utility should accept standard compression flags (`-c` for standard output, `-d` for decompress.)
* schema: In case the snapshots should be cleaned, this is the schema the manager will use to clean.
* local_schema: For local snapshot cleaning/aging when dataset is receptical for remote source when snapshots are pulled
* remote_schema: For remote snapshot cleaning/aging when remote target is receptical for backup when snapshots are pushed
* remote2_schema: For remote snapshot cleaning/aging when remote target is receptical for backup when snapshots are pushed
* preexec: A command that will be executed, before snapshot/replication. Should be omitted if nothing should be executed
* postexec: A command that will be executed, after snapshot/replication,  but before the cleanup. Should be omitted if nothing should be executed
* clean_all: Clean/age all snapshots in dataset - default is False - ie zsnapd only
* local_clean_all: Setting for local dataset when replicating source is remote
* remote_clean_all: Setting for remove dataset when replicating source is local
* remote2_clean_all: Setting for remove dataset when replicating source is local
* all_snapshots: Replicate all snapshots in dataset - Default is True - ie all snapshots in dataset
* replicate_all: Deprecated.  Use all_snapshots
* log_commands: Per dataset log all commands executed for the dataset to DEBUG.  For checking what the program is doing exactly, helpful for auditing and security. 

Naming convention
-----------------

This script's snapshot will always given a timestamp (format yyyymmddhhmm) as name. For pool/tank an
example snapshot name could be pool/tank@201312311323.  The daemon is still compatible with the olderyyyymmdd 
snapshot aging convention, and will replicate and age them.  (Internally, the snapshot 'handles' are now created from
the snapshot creation time (using Unix timestamp seconds) - this means that manual snapshot names are covered too.)

All snapshots are currently used for replication (both snapshots taken by the script as well as snapshots taken by
other means (other scripts or manually), regardless of their name.

However, the script will only clean snapshots with the timestamp naming convention. In case you don't want a snapshot to
be cleaned by the script, just make sure it has any other name, not matching this convention.

Buckets
-------

The system will use "buckets" to apply the GFS schema.
From every bucket, the oldest snapshot will be kept. At any given time the script is
executed, it will place the snapshots in their buckets, and then clean out all buckets.

Bucket schema
-------------

For example, the schema '7d3w11m4y' means:

* 7 daily buckets (starting from today)
* 3 weekly buckets (7 days a week)
* 11 monthly buckets (30 days a month)
* 4 yearly buckets (12 * 30 days a year)

This wraps up to 5 years (where a year is 12 * 30 days - so not mapping to a real year)

Other schema's are possible. One could for example only be intrested in keeping only the
snapshots for last week, in which the schema '7d0w0m0y' would be given. Any combination is possible.

Since from each bucket, the oldest snapshot is kept, snapshots will seem to "roll"
trough the buckets.

The number of 'keep' days before aging can be given, as well as hourly buckets before the days kick in, 
ie '2k24h7d3w11m4y':

* 2 keep days starting from midnight, no snapshots deleted
* 24 hourly buckets starting from midnight in 2 days
* 7 daily buckets
* 3 weekly buckets (7 days a week)
* 11 monthly buckets (30 days a month)
* 4 yearly buckets (12 * 30 days a year)

Remote Execution Security
-------------------------

Sudo with a backup user on the remote was considered, but after reviewing the
sshd ForceCommand mechanism for remote excution, this was chosen as far easier
and superior. Thus, zsnapd-rcmd, the remote sshd command checker for ZFS
Snapshot Daemon

zsnapd-rcmd is the security plugin command for sshd that implements ForceCommand
functionality, or the command functionality in the .ssh/authorized_keys file
(See the sshd_config(8) and sshd(8) man pages respectively).

It executes commands from the SSH_ORIGINAL_COMMAND variable after checking
them against a list of configured regular expressions. 

Edit the zsnapd-rcmd.conf files in /etc/zsnapd, to set up the check regexps
for the remote preexec, postexec, and replicate_postexec commands.  Settings
are also available for 10 extra remote commands, labeled rcmd_aux0 - rcmd_aux9

Read the sshd(8) manpage on the ForceCommand setting, and the sshd(8) manpage
on the /root/.ssh/authorized_keys file, command entry for the remote pub key
for zsnapd access.

Example .ssh/authorized_keys entry (single line - unline wrap it):

no-pty,no-agent-forwarding,no-X11-forwarding,no-port-forwarding,
command="/usr/sbin/zsnapd-rcmd" ssh-rsa AAAABBBBBBBBCCCCCCCCCCCCC
DDDDDDDD== root@blah.org

Hint: command line arguments can be given, such as a different config file,
and debug level:

no-pty,no-agent-forwarding,no-X11-forwarding,no-port-forwarding,
command="/usr/sbin/zsnapd-rcmd -c /etc/zsnapd/my-rcmd.conf --debug 3" 
ssh-rsa AAAABBBBBBBBCCCCCCCCCCCCCDDDDDDDD== root@blah.org

Examples
--------

The examples directory contains 3 example configuration files, almost identical as my own 'production' setup.

* A non-ZFS device (router), rsyncing its filesystem to an NFS shared dataset.
* A laptop, having a single root ZFS setup, containing 2 normal filesystems and a ZFS dataset
* A local NAS with lots of data and the replication target of most systems
* A remote NAS (used as normal NAS by these people) used with two-way replication as offsite backup setup.

Dependencies
------------

This python program/script has a few dependencies. When using the Archlinux AUR, these will be installed automatically.

* zfs
* python3
* openssh
* mbuffer
* python3-magcode-core >= 1.5.4 - on pypi.org
* python3-psutil
* python3-setproctitle

Logging
-------

The script is logging into systemd journals, and /var/log/syslog

License
-------

This program/script is licensed under MIT, which basically means you can do anything you want with it. You can find
the license text in the 'LICENSE' file.

If you like the software or if you're using it, feel free to leave a star as a toke of appreciation.

Warning
-------

As with any script deleting snapshots, use with caution. Make sure to test the script on
a dummy dataset first when you use it directly from the repo. This to ensure no unexpected things will happen.

The releases should be working fine, as I use these on my own environment, and for customers.

In case you find a bug, feel free to create a bugreport and/or fork and send a pull-request
in case you fixed the bug yourself.

Packages
--------

ZFS Snapshot Manager is available in the following distributions:

* ArchLinux: https://aur.archlinux.org/packages/zfs-snap-manager (AUR)
  * The PKGBUILD and install scripts are now available through the AUR git repo

zsnapd is available in following distributions:

* Debian: http://packages.debian.org as part of the main repostitory
* Ubuntu (eventually)

ZFS
---

From Wikipedia:

    ZFS is a combined file system and logical volume manager designed by Sun Microsystems.
    The features of ZFS include protection against data corruption, support for high storage capacities,
    efficient data compression, integration of the concepts of filesystem and volume management, snapshots
    and copy-on-write clones, continuous integrity checking and automatic repair, RAID-Z and native NFSv4 ACLs.

    ZFS was originally implemented as open-source software, licensed under the Common Development and
    Distribution License (CDDL). The ZFS name is registered as a trademark of Oracle Corporation.

ZFS Snapshot Manager and zsnapd are standalone projects, and is not affiliated with ZFS or Oracle Corporation.


debian/changelog
zsnapd (0.8.11f-1) unstable; urgency=medium

  * Update README.md for trigger imporvements and receive mountpoint work.

 -- Matthew Grant <matt@mattgrant.net.nz>  Fri, 06 Mar 2020 09:09:02 +1300

zsnapd (0.8.11e-1) unstable; urgency=medium

  * Read /proc/self/mounts to determine default mount info, pass if
    not available.

 -- Matthew Grant <matt@mattgrant.net.nz>  Fri, 06 Mar 2020 08:39:52 +1300

zsnapd (0.8.11d-1) unstable; urgency=medium

  * Fix bug  default intialisation receive_mountpoint

 -- Matthew Grant <matt@mattgrant.net.nz>  Thu, 05 Mar 2020 12:47:19 +1300

zsnapd (0.8.11c-1) unstable; urgency=medium

  * Make --do-trigger use dateset.conf do_trigger flag for cleaner
    integration with devops tools like ansible.

 -- Matthew Grant <matt@mattgrant.net.nz>  Thu, 05 Mar 2020 11:36:55 +1300

zsnapd (0.8.11b-1) unstable; urgency=medium

  * Better zsnapd-trigger messages, do_trigger flag

 -- Matthew Grant <matt@mattgrant.net.nz>  Thu, 05 Mar 2020 09:44:34 +1300

zsnapd (0.8.11a-1) unstable; urgency=medium

  * Append names and receive mountpoint work

 -- Matthew Grant <matt@mattgrant.net.nz>  Mon, 02 Mar 2020 12:22:32 +1300

zsnapd (0.8.10b-1) unstable; urgency=medium

  * Update documentation for new umount option

 -- Matthew Grant <matt@mattgrant.net.nz>  Sun, 19 Jan 2020 09:09:06 +1300

zsnapd (0.8.10a-1) unstable; urgency=medium

  * Add umount arg support to zfs recv

 -- Matthew Grant <matt@mattgrant.net.nz>  Sun, 19 Jan 2020 08:50:46 +1300

zsnapd (0.8.10-1) unstable; urgency=medium
 
  * release 0.8.10
  * -x mountpoint support
  * -w send/recv raw transfer support
  * -s restore save point and auto - resume send/recv support
  * Cleaned ZFS hold manipulation
  * Better per dataset listing of snapshots etc.
  * Set default -x mountpoint for full replication etc

 -- Matthew Grant <matt@mattgrant.net.nz>  Fri, 17 Jan 2020 20:37:37 +1300

zsnapd (0.8.9o-1~1.gbp0e208j) unstable; urgency=medium

  * Fix rcmd_zfs_holds regex to actually work

 -- Matthew Grant <matt@mattgrant.net.nz>  Wed, 15 Jan 2020 22:12:29 +1300

zsnapd (0.8.9n-1~1.gbp0e208i) unstable; urgency=medium

  * Correct bug in zsnapd-rcmd misnamed config item
  * Fix missing line in zsnapd-rcmd globals

 -- Matthew Grant <matt@mattgrant.net.nz>  Wed, 15 Jan 2020 22:01:31 +1300

zsnapd (0.8.9m-1~1.gbp0e208h) unstable; urgency=medium

  * Resume zfs send/receive implemented. v1.
  * Finish updating log messages
  * Implement get_snapshots2, more efficient and clear

 -- Matthew Grant <matt@mattgrant.net.nz>  Wed, 15 Jan 2020 21:53:14 +1300

zsnapd (0.8.9j-1~1.gbp0e208g) unstable; urgency=medium

  * Zsnapd rcmd rcmd_zfs_holds bug fixes

 -- Matthew Grant <matt@mattgrant.net.nz>  Wed, 15 Jan 2020 12:02:35 +1300

zsnapd (0.8.9i-1~1.gbp0e208f) unstable; urgency=medium

  * Bug fix for ValueError in Manager.new_hold()

 -- Matthew Grant <matt@mattgrant.net.nz>  Wed, 15 Jan 2020 11:28:05 +1300

zsnapd (0.8.9h-1~1.gbp0e208e) unstable; urgency=medium

  * WIP - resume token usage
  * WIP - New hold managment commands

 -- Matthew Grant <matt@mattgrant.net.nz>  Wed, 15 Jan 2020 11:24:16 +1300

zsnapd (0.8.9f-1~1.gbp0e208d) unstable; urgency=medium

  * Merge v0.8.9f test release.
  * Interrupted zfs send/recv support - resume commands

 -- Matthew Grant <matt@mattgrant.net.nz>  Tue, 14 Jan 2020 23:11:46 +1300

zsnapd (0.8.9e-1~1.gbp0e208c) unstable; urgency=medium

  ** SNAPSHOT build @0e208c8213797934c8de18863c3f45f534b4ce69 **

  * Add send raw -w flag to support encryption
  * Update zsnapd-trigger help message
  * Add failed transfer save points 1

 -- Matthew Grant <matt@mattgrant.net.nz>  Wed, 08 Jan 2020 21:05:15 +1300

zsnapd (0.8.9d-2) unstable; urgency=medium

  * Correct distribution entries to unstable only.

 -- Matthew Grant <matt@mattgrant.net.nz>  Thu, 02 Jan 2020 13:20:50 +1300

zsnapd (0.8.9d-1) unstable; urgency=medium

  * Move package to contrib section of archive. (Closes: #919595)

 -- Matthew Grant <matt@mattgrant.net.nz>  Thu, 02 Jan 2020 13:18:27 +1300

zsnapd (0.8.9d-1~1.gbp9bc387) unstable; urgency=medium

  * Log message updates and no error on existing hold

 -- Matthew Grant <matt@mattgrant.net.nz>  Wed, 10 Apr 2019 11:20:57 +1200

zsnapd (0.8.9c-1~1.gbp9bc386) unstable; urgency=medium

  * Update changelog for 0.8.9b-1~1.gbp9bc385 release
  * Bug fixes for replicate2

 -- Matthew Grant <matt@mattgrant.net.nz>  Tue, 09 Apr 2019 14:24:43 +1200

zsnapd (0.8.9b-1~1.gbp9bc385) unstable; urgency=medium

  ** SNAPSHOT build @9bc38552f1c45b2838643856214302a3bd78c868 **

  * WIP - initial replicate2 config reading
  * replicate2 2nd dataset target support

 -- Matthew Grant <matt@mattgrant.net.nz>  Tue, 09 Apr 2019 13:47:05 +1200

zsnapd (0.8.9a-1) unstable; urgency=medium

  * Version 0.8.9a
  * correct a config syntax log msg

 -- Matthew Grant <matt@mattgrant.net.nz>  Thu, 21 Feb 2019 23:31:24 +1300

zsnapd (0.8.9-1) unstable; urgency=medium

  * Fix Cleaing start log message to match standard format
  * Add ability to read mountpoint property from ZFS
  * trigger allowed in comma time spec
  * {template} key in config for time string
  * Version 0.8.9, documentation updated

 -- Matthew Grant <matt@mattgrant.net.nz>  Thu, 21 Feb 2019 22:21:13 +1300

zsnapd (0.8.8-1) unstable; urgency=medium

  * Release 0.8.8
  * Fix cleaning bugs if no snapshotting

 -- Matthew Grant <matt@mattgrant.net.nz>  Mon, 18 Feb 2019 19:29:40 +1300

zsnapd (0.8.7-1) unstable; urgency=medium

  * Version 0.8.7
  * Update TODO list for future work on event queueing
  * Fix time interval calculation and parsing

 -- Matthew Grant <matt@mattgrant.net.nz>  Sun, 10 Feb 2019 14:24:34 +1300

zsnapd (0.8.5-1) unstable; urgency=medium

  * Version 0.8.5 release - fixes a few serious time logic issues
  * Make snapshot creation time per dataset
  * Implement startup_hysteresis_time for MeterTime

 -- Matthew Grant <matt@mattgrant.net.nz>  Fri, 01 Feb 2019 15:09:12 +1300

zsnapd (0.8.4-1) unstable; urgency=medium


scripts/globals_.py
"""
Globals file for zsnapd
"""

from magcode.core.globals_ import settings

# Constants for use in program
CLEANER_REGEX = r'^((?P<keep>[0-9]+)k){0,1}((?P<hours>[0-9]+)h){0,1}(?P<days>[0-9]+)d(?P<weeks>[0-9]+)w(?P<months>[0-9]+)m(?P<years>[0-9]+)y$'
SNAPSHOTNAME_REGEX = r'^(\d{4})(1[0-2]|0[1-9])(0[1-9]|[1-2]\d|3[0-1])(([0-1]\d|2[0-3])([0-5]\d)){0,1}$'
SNAPSHOTNAME_FMTSPEC = '%Y%m%d%H%M'
TRIGGER_FILENAME = '.trigger'
DEFAULT_BUFFER_SIZE = '512M'

# settings for where files are
settings['config_dir'] = '/etc/zsnapd'
settings['log_dir'] = '/var/log/zsnapd'
settings['run_dir'] = '/run'
settings['config_file'] = settings['config_dir'] + '/' + 'process.conf'
# Zsnapd only uses one daemon
settings['pid_file'] = settings['run_dir'] + '/' + 'zsnapd.pid'
#settings['log_file'] = settings['log_dir'] \
#        + '/' + settings['process_name'] + '.log'
settings['log_file'] = ''
settings['panic_log'] = settings['log_dir'] \
        + '/' + settings['process_name'] + '-panic.log'
settings['syslog_facility'] = ''

# zsnapd.py
# Dataset config file
settings['dataset_config_file'] = settings['config_dir'] \
        + '/' + 'dataset.conf'
settings['dataset_config_dir'] = settings['config_dir'] \
        + '/' + 'dataset.conf.d'
# Template config file
settings['template_config_file'] = settings['config_dir'] \
        + '/' + 'template.conf'
settings['template_config_dir'] = settings['config_dir'] \
        + '/' + 'template.conf.d'
# Print debug mark
settings['debug_mark'] = False
# Number of seconds we wait while looping in main loop...
settings['sleep_time'] = 300 # seconds
settings['debug_sleep_time'] = 15 # seconds
settings['startup_hysteresis_time'] = 15 # seconds
settings['connect_retry_wait'] = 3 # seconds

settings['zfs_proc_not_mounts'] = ('/var/lib/lxd/devices',)
def read_proc_mounts():
    zfs_mnts = {}
    try:
        for mnt in open('/proc/self/mounts'):
            mnt = mnt.split(' ')
            if (len(mnt) < 3):
                continue
            if (mnt[2] != 'zfs'):
                continue
            if (mnt[1].startswith(settings['zfs_proc_not_mounts'])):
                continue
            zfs_mnts[mnt[0]] = mnt[1]
    except:
        pass
    return zfs_mnts
settings['zfs_proc_mounts'] = read_proc_mounts()
 
